Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine+Demand.swift

```diff
@@ -0,0 +1,149 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+extension HTTPRequestStateMachine {
+    /// A sub state for receiving a response events. Stores whether the consumer has either signaled demand and whether the
+    /// channel has issued `read` events.
+    struct ResponseStreamState {
+        private enum State {
+            /// The state machines expects further writes to `channelRead`. The writes are appended to the buffer.
+            case waitingForBytes(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `demandMoreResponseBodyParts` or `read`. The buffer is
+            /// empty. It is preserved for performance reasons.
+            case waitingForReadOrDemand(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `read`. The buffer is empty. It is preserved for performance reasons.
+            case waitingForRead(CircularBuffer<ByteBuffer>)
+            /// The state machines expects a call to `demandMoreResponseBodyParts`. The buffer is empty. It is
+            /// preserved for performance reasons.
+            case waitingForDemand(CircularBuffer<ByteBuffer>)
+
+            case modifying
+        }
+
+        enum Action {
+            case read
+            case wait
+        }
+
+        private var state: State
+
+        init(expectingBody: Bool) {
+            self.state = .waitingForBytes(CircularBuffer(initialCapacity: expectingBody ? 16 : 0))
```

Comment:
Given that there is no such thing as an empty circular buffer (`initialCapacity: 0` still allocates) we may as well skip the branch and just always set `16`.